clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java

```diff
@@ -151,236 +104,142 @@ else if (clientAuthConfig.equals("requested"))
+            if (keystoreVerifiableUsingTruststore) {
+                if (sslEngineBuilder.truststore() != null || sslEngineBuilder.keystore() != null) {
+                    SslEngineValidator.validate(sslEngineBuilder, newSslEngineBuilder);
+                }
             }
         } catch (Exception e) {
-            log.debug("Validation of dynamic config update of SSL keystore/truststore failed", e);
-            throw new ConfigException("Validation of dynamic config update of SSL keystore/truststore failed: " + e);
+            log.info("Unable to validate proposed new SslFactory configuration.", e);
+            throw new ConfigException("Unable to reconfigure SslFactory: " + e);
         }
     }
 
     @Override
-    public void reconfigure(Map<String, ?> configs) throws KafkaException {
-        SecurityStore newKeystore = maybeCreateNewKeystore(configs);
-        SecurityStore newTruststore = maybeCreateNewTruststore(configs);
-        if (newKeystore != null || newTruststore != null) {
-            try {
-                SecurityStore keystore = newKeystore != null ? newKeystore : this.keystore;
-                SecurityStore truststore = newTruststore != null ? newTruststore : this.truststore;
-                this.sslContext = createSSLContext(keystore, truststore);
-                log.info("Created new SSL context with keystore {} truststore {}", keystore, truststore);
-                this.keystore = keystore;
-                this.truststore = truststore;
-            } catch (Exception e) {
-                log.debug("Reconfiguration of SSL keystore/truststore failed", e);
-                throw new ConfigException("Reconfiguration of SSL keystore/truststore failed: " + e);
-            }
+    public void reconfigure(Map<String, ?> newConfigs) throws KafkaException {
+        if (sslEngineBuilder == null) {
+            throw new IllegalStateException("SslFactory has not been configured.");
         }
-    }
-
-    private SecurityStore maybeCreateNewKeystore(Map<String, ?> configs) {
-        boolean keystoreChanged  = false;
-        if (keystore != null) {
-            keystoreChanged = !Objects.equals(configs.get(SslConfigs.SSL_KEYSTORE_TYPE_CONFIG), keystore.type) ||
-                    !Objects.equals(configs.get(SslConfigs.SSL_KEYSTORE_LOCATION_CONFIG), keystore.path) ||
-                    !Objects.equals(configs.get(SslConfigs.SSL_KEYSTORE_PASSWORD_CONFIG), keystore.password) ||
-                    !Objects.equals(configs.get(SslConfigs.SSL_KEY_PASSWORD_CONFIG), keystore.keyPassword);
-            if (!keystoreChanged) {
-                keystoreChanged = keystore.modified();
-            }
+        Map<String, Object> nextConfigs = new HashMap<>(sslEngineBuilder.configs());
+        copyMapEntries(nextConfigs, newConfigs, SslConfigs.RECONFIGURABLE_CONFIGS);
+        if (clientAuthConfigOverride != null) {
+            nextConfigs.put(BrokerSecurityConfigs.SSL_CLIENT_AUTH_CONFIG, clientAuthConfigOverride);
```

Comment:
Same as in validate, why the special handling? Isn't this included under `sslEngineBuilder.configs()`?

Reply:
same issue as above -- the override should overwrite whatever the user supplied